visionmetrics/detection/confusion_matrix.py

```diff
@@ -0,0 +1,140 @@
+import torch
+from torchmetrics import Metric
+
+
+class DetectionConfusionMatrix(Metric):
+    """
+        Calcuates confusion matrix in the context of object detection.
+
+        Parameters
+        ----------
+        predictions : list of lists
+            Each list contains list of bounding box predictions for an image in the format [class_id, score, xmin, ymin, xmax, ymax]
+        targets : list of lists
+            Each list contains ground truth bounding boxes for an image in the format [class_id, xmin, ymin, xmax, ymax]
+        iou_threshold : float
+            IoU threshold to consider a detection as True Positive
+
+        Example:
+        predictions = [[[0, 0.9, 10, 20, 50, 100], [1, 0.8, 30, 40, 80, 120]], [[1, 0.7, 20, 30, 60, 90]]]
+        targets = [[[0, 10, 20, 50, 100], [1, 30, 40, 80, 120]], [[1, 20, 30, 60, 90]]]
+
+        Returns
+        -------
+        confusion_matrix : dict
+            Dictionary containing the counts of TP, FP, and FN, as well as details for FP reasons
+        """
+
+    def __init__(self, iou_threshold=0.5, **kwargs):
+        super().__init__(**kwargs)
+        self.iou_threshold = iou_threshold
+        self.add_state("tp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fn", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_wrong_class", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_low_iou", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("fp_due_to_no_corresponding_gt_box", default=torch.tensor(0), dist_reduce_fx="sum")
+        self.add_state("confusion_matrix", default=[], dist_reduce_fx=None)
+
+    def update(self, predictions, targets):
+        if len(predictions) != len(targets):
+            raise ValueError("Number of predictions and targets should be the same.")
+        if not isinstance(predictions[0][0], list):
+            raise ValueError(f"Expected predictions to be a list of lists, got a list of {type(predictions[0][0])}")
+
+        self.confusion_matrix.append(self._update_confusion_matrix(predictions, targets, self.iou_threshold))
```

Comment:
maybe we don't need to construct the dictionary here, just update the tp*, fp*. then create a dictionary in `compute()` method?

Reply:
Great suggestion! Updated.